server/service/apple_mdm.go

```diff
@@ -3864,125 +3866,186 @@ func uploadABMTokenEndpoint(ctx context.Context, request interface{}, svc fleet.
+// Update ABM token teams endpoint
+////////////////////////////////////////////////////////////////////////////////
+
+type updateABMTokenTeamsRequest struct {
+	TokenID      uint  `url:"id"`
+	MacOSTeamID  *uint `json:"macos_team_id"`
+	IOSTeamID    *uint `json:"ios_team_id"`
+	IPadOSTeamID *uint `json:"ipados_team_id"`
+}
+
+type updateABMTokenTeamsResponse struct {
+	ABMToken *fleet.ABMToken `json:"abm_token,omitempty"`
+	Err      error           `json:"error,omitempty"`
+}
+
+func (r updateABMTokenTeamsResponse) error() error { return r.Err }
+
+func updateABMTokenTeamsEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (errorer, error) {
+	req := request.(*updateABMTokenTeamsRequest)
+
+	tok, err := svc.UpdateABMTokenTeams(ctx, req.TokenID, req.MacOSTeamID, req.IOSTeamID, req.IPadOSTeamID)
 	if err != nil {
-		return ctxerr.Wrap(ctx, err, "retrieving app config")
+		return &updateABMTokenTeamsResponse{Err: err}, nil
 	}
 
-	appCfg.MDM.AppleBMEnabledAndConfigured = true
+	return &updateABMTokenTeamsResponse{ABMToken: tok}, nil
+}
+
+func (svc *Service) UpdateABMTokenTeams(ctx context.Context, tokenID uint, macOSTeamID, iOSTeamID, iPadOSTeamID *uint) (*fleet.ABMToken, error) {
+	// skipauth: No authorization check needed due to implementation returning
+	// only license error.
+	svc.authz.SkipAuthorization(ctx)
 
-	return svc.ds.SaveAppConfig(ctx, appCfg)
+	return nil, fleet.ErrMissingLicense
 }
 
 ////////////////////////////////////////////////////////////////////////////////
-// Disable ABM endpoint
+// Renew ABM token endpoint
 ////////////////////////////////////////////////////////////////////////////////
 
-type disableABMResponse struct {
-	Err error `json:"error,omitempty"`
+type renewABMTokenRequest struct {
+	TokenID uint `url:"id"`
+	Token   *multipart.FileHeader
 }
 
-func (r disableABMResponse) error() error { return r.Err }
-func (r disableABMResponse) Status() int  { return http.StatusNoContent }
+func (renewABMTokenRequest) DecodeRequest(ctx context.Context, r *http.Request) (interface{}, error) {
+	err := r.ParseMultipartForm(512 * units.MiB)
+	if err != nil {
+		return nil, &fleet.BadRequestError{
+			Message:     "failed to parse multipart form",
+			InternalErr: err,
+		}
+	}
 
-func disableABMEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (errorer, error) {
-	if err := svc.DisableABM(ctx); err != nil {
-		return disableABMResponse{Err: err}, nil
+	token, ok := r.MultipartForm.File["token"]
+	if !ok || len(token) < 1 {
+		return nil, &fleet.BadRequestError{Message: "no file headers for token"}
 	}
 
-	return disableABMResponse{}, nil
-}
+	// because we are in this method, we know that the path has 6 parts, e.g:
+	// /api/latest/fleet/abm_tokens/19/renew
 
-func (svc *Service) DisableABM(ctx context.Context) error {
-	if err := svc.authz.Authorize(ctx, &fleet.AppleBM{}, fleet.ActionWrite); err != nil {
-		return err
+	pathParts := strings.Split(r.URL.Path, "/")
```

Comment:
This endpoint seems to be unique in that it passes an ID in the URL but also has a file in the form body. I didn't see another example of this, so I had to roll my own URL parsing here. LMK if that's not right!

Reply:
Ooh, true that this is a special case. We should really add support for the multipart/file upload decoding as part of our "framework" stuff at this point, but what you should be able to do here is replicate what it does in the `makeDecoder` implementation, you should be able to call this: https://github.com/fleetdm/fleet/blob/374f5496a67d5e6c72a2bbe654f1fe31dc568d82/server/service/transport.go#L90-L101 with the name of the url parameter (`id` in this case).

Reply:
oh great idea!